setting interface from nested func call

### The following program `sample.go` triggers an unexpected result

```go
package main
                                                                                
type Pooler interface {                                                         
    Get() string                                                                
}                                                                               
                                                                                
type baseClient struct {                                                        
    connPool Pooler                                                             
}                                                                               
                                                                                
type connPool struct {                                                          
    name string                                                                 
}                                                                               
                                                                                
func (c *connPool) Get() string { return c.name }                               
                                                                                
func newBaseClient(p Pooler) *baseClient {                                      
    return &baseClient{connPool: p}                                             
}                                                                               
                                                                                
func newConnPool() *connPool { return &connPool{name: "connPool"} }             

func main() {                                                                   
    b := newBaseClient(newConnPool())                                           
    println(b.connPool.(*connPool).name)                                        
}
```


### Expected result

```console
$ go run ./sample.go
connPool
```


### Got

```console
$ yaegi ./sample.go
/Users/marc/yt/call1.go:24:21: panic
run: reflect.Set: value of type *struct { Xname string } is not assignable to type interp.valueInterface
goroutine 1 [running]:
runtime/debug.Stack()
	/Users/marc/sdk/go1.17.3/src/runtime/debug/stack.go:24 +0x88
github.com/traefik/yaegi/interp.(*Interpreter).Execute.func1(0x140001bbae0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/program.go:131 +0x8c
panic({0x104f8ecc0, 0x140001a38c0})
	/Users/marc/sdk/go1.17.3/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp.runCfg.func1(0x140000d0bb0, 0x140004bec60, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:193 +0x164
panic({0x104f8ecc0, 0x140001a38c0})
	/Users/marc/sdk/go1.17.3/src/runtime/panic.go:1038 +0x21c
reflect.Value.assignTo({0x140001ceb40, 0x14000010b50, 0x196}, {0x104da9463, 0xb}, 0x1050075e0, 0x0)
	/Users/marc/sdk/go1.17.3/src/reflect/value.go:2797 +0x2cc
reflect.Value.Set({0x1050075e0, 0x14000033460, 0x199}, {0x140001ceb40, 0x14000010b50, 0x196})
	/Users/marc/sdk/go1.17.3/src/reflect/value.go:1905 +0xa4
github.com/traefik/yaegi/interp.call.func7(0x140000d0bb0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:1364 +0xb4c
github.com/traefik/yaegi/interp.runCfg(0x140004bec60, 0x140000d0bb0, 0x140004be120, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:201 +0x268
github.com/traefik/yaegi/interp.(*Interpreter).run(0x140000f9b00, 0x140004be120, 0x140000d0a50)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:120 +0x3d4
github.com/traefik/yaegi/interp.(*Interpreter).Execute(0x140000f9b00, 0x140004c5710)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/program.go:157 +0x1e0
github.com/traefik/yaegi/interp.(*Interpreter).eval(0x140000f9b00, {0x140002ea780, 0x1c3}, {0x16b8eb728, 0x17}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:615 +0x98
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0x140000f9b00, {0x16b8eb728, 0x17})
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:508 +0xe8
main.runFile(0x140000f9b00, {0x16b8eb728, 0x17}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:154 +0xb0
main.run({0x14000032050, 0x1, 0x1})
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:117 +0xe14
main.main()
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x384
```


### Yaegi Version

v0.11.0

### Additional Notes

_No response_

--- a/_test/issue-1320.go
+++ b/_test/issue-1320.go
@@ -1,0 +1,29 @@
+package main
+
+type Pooler interface {
+	Get() string
+}
+
+type baseClient struct {
+	connPool Pooler
+}
+
+type connPool struct {
+	name string
+}
+
+func (c *connPool) Get() string { return c.name }
+
+func newBaseClient(i int, p Pooler) *baseClient {
+	return &baseClient{connPool: p}
+}
+
+func newConnPool() *connPool { return &connPool{name: "connPool"} }
+
+func main() {
+	b := newBaseClient(0, newConnPool())
+	println(b.connPool.(*connPool).name)
+}
+
+// Output:
+// connPool

--- a/interp/run.go
+++ b/interp/run.go
@@ -1125,27 +1125,28 @@ func methodByName(value reflect.Value, name string) reflect.Value {
 func call(n *node) {
 	goroutine := n.anc.kind == goStmt
 	var method bool
-	value := genValue(n.child[0])
+	c0 := n.child[0]
+	value := genValue(c0)
 	var values []func(*frame) reflect.Value
 
 	recvIndexLater := false
 	switch {
-	case n.child[0].recv != nil:
+	case c0.recv != nil:
 		// Compute method receiver value.
-		values = append(values, genValueRecv(n.child[0]))
+		values = append(values, genValueRecv(c0))
 		method = true
-	case len(n.child[0].child) > 0 && n.child[0].child[0].typ != nil && isInterfaceSrc(n.child[0].child[0].typ):
+	case len(c0.child) > 0 && c0.child[0].typ != nil && isInterfaceSrc(c0.child[0].typ):
 		recvIndexLater = true
-		values = append(values, genValueBinRecv(n.child[0], &receiver{node: n.child[0].child[0]}))
+		values = append(values, genValueBinRecv(c0, &receiver{node: c0.child[0]}))
 		value = genValueBinMethodOnInterface(n, value)
 		method = true
-	case n.child[0].action == aMethod:
+	case c0.action == aMethod:
 		// Add a place holder for interface method receiver.
 		values = append(values, nil)
 		method = true
 	}
 
-	numRet := len(n.child[0].typ.ret)
+	numRet := len(c0.typ.ret)
 	variadic := variadicPos(n)
 	child := n.child[1:]
 	tnext := getExec(n.tnext)
@@ -1153,27 +1154,40 @@ func call(n *node) {
 
 	// Compute input argument value functions.
 	for i, c := range child {
+		var arg *itype
+		if variadic >= 0 && i >= variadic {
+			arg = c0.typ.arg[variadic].val
+		} else {
+			arg = c0.typ.arg[i]
+		}
 		switch {
 		case isBinCall(c, c.scope):
 			// Handle nested function calls: pass returned values as arguments.
 			numOut := c.child[0].typ.rtype.NumOut()
 			for j := 0; j < numOut; j++ {
 				ind := c.findex + j
-				values = append(values, func(f *frame) reflect.Value { return f.data[ind] })
+				if !isInterfaceSrc(arg) || isEmptyInterface(arg) {
+					values = append(values, func(f *frame) reflect.Value { return f.data[ind] })
+					continue
+				}
+				values = append(values, func(f *frame) reflect.Value {
+					return reflect.ValueOf(valueInterface{value: f.data[ind]})
+				})
 			}
 		case isRegularCall(c):
 			// Arguments are return values of a nested function call.
-			for j := range c.child[0].typ.ret {
+			cc0 := c.child[0]
+			for j := range cc0.typ.ret {
 				ind := c.findex + j
-				values = append(values, func(f *frame) reflect.Value { return f.data[ind] })
+				if !isInterfaceSrc(arg) || isEmptyInterface(arg) {
+					values = append(values, func(f *frame) reflect.Value { return f.data[ind] })
+					continue
+				}
+				values = append(values, func(f *frame) reflect.Value {
+					return reflect.ValueOf(valueInterface{value: f.data[ind]})
+				})
 			}
 		default:
-			var arg *itype
-			if variadic >= 0 && i >= variadic {
-				arg = n.child[0].typ.arg[variadic].val
-			} else {
-				arg = n.child[0].typ.arg[i]
-			}
 			if c.kind == basicLit || c.rval.IsValid() {
 				argType := arg.TypeOf()
 				convertLiteralValue(c, argType)
@@ -1193,7 +1207,7 @@ func call(n *node) {
 	}
 
 	// Compute output argument value functions.
-	rtypes := n.child[0].typ.ret
+	rtypes := c0.typ.ret
 	rvalues := make([]func(*frame) reflect.Value, len(rtypes))
 	switch n.anc.kind {
 	case defineXStmt, assignXStmt:
@@ -1228,7 +1242,7 @@ func call(n *node) {
 
 	if n.anc.kind == deferStmt {
 		// Store function call in frame for deferred execution.
-		value = genFunctionWrapper(n.child[0])
+		value = genFunctionWrapper(c0)
 		if method {
 			// The receiver is already passed in the function wrapper, skip it.
 			values = values[1:]

--- a/interp/run.go
+++ b/interp/run.go
@@ -1184,7 +1184,7 @@ func call(n *node) {
 					continue
 				}
 				values = append(values, func(f *frame) reflect.Value {
-					return reflect.ValueOf(valueInterface{value: f.data[ind]})
+					return reflect.ValueOf(valueInterface{node: cc0.typ.ret[j].node, value: f.data[ind]})
 				})
 			}
 		default:

--- a/interp/typecheck.go
+++ b/interp/typecheck.go
@@ -850,7 +850,7 @@ func (check typecheck) builtin(name string, n *node, child []*node, ellipsis boo
 			return params[0].nod.cfgErrorf("first argument to delete must be map; have %s", typ.id())
 		}
 		ktyp := params[1].Type()
-		if !ktyp.assignableTo(typ.key) {
+		if typ.key != nil && !ktyp.assignableTo(typ.key) {
 			return params[1].nod.cfgErrorf("cannot use %s as type %s in delete", ktyp.id(), typ.key.id())
 		}
 	case bltnMake:

